### PR TITLE
[SFT-671] - Support ability to encrypt/decrypt submission data

### PR DESCRIPTION
**Pull request by seandelaney**
- Added new field property to encrypt field data.
- Added checks to show new encrypt field data property for Lite and Pro editions.
- Added new Encryption helper.
- Updated `Submission` element to encrypt submission fields on `beforeSave` if field can be encrypted and setting enabled.
- Updated `SubmissionQuery` to trigger decrypt submission fields on `afterPopulateElement` if field can be decrypted.
- Added new `EncryptionInterface` and `EncryptionTrait` to control which fields can be encrypted/decrypted.
- Fixed exporting submission data.
- Added new Input Field Type to better manage form controls

### CODE REVIEW

**Review by kjmartens — DISMISSED** [review]
When I try viewing the submission data inside the CP, front end and "Updating Search Indexes" in Craft Queue, I get an error:
`base64_decode(): Argument #1 ($string) must be of type string, array given`

When attempting to submit a multipage form (not sure if it's the fact that it's multipage or something else), the form fails at the final submit and has this error:
```
Exception 'TypeError' with message 'openssl_encrypt(): Argument #1 ($data) must be of type string, array given'

in /YOINK/freeform5.c4/vendor/yiisoft/yii2/base/Security.php:218
```

On a single page form it appeared to work ok for submit.

File: packages/plugin/src/Bundles/Fields/Types/FieldTypesProvider.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Bundles/Fields/Types/FieldTypesProvider.php]: Why did you remove this?
There are times when no default section is preconfigured.

File: packages/plugin/src/Bundles/Submissions/CipherFieldsBundle.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Bundles/Submissions/CipherFieldsBundle.php]: With PHP8.0 you can use constructor arguments to define object properties. To avoid having to write boilerplate.

```suggestion
class CipherFieldsBundle extends FeatureBundle
{
    public function __construct(private CipherHelper $cipherHelper)
    {
        Event::on(
            Submission::class,
            Submission::EVENT_ENCRYPT_FIELDS,
            [$this, 'encryptFields']
        );

        Event::on(
            Submission::class,
            Submission::EVENT_DECRYPT_FIELDS,
            [$this, 'decryptFields']
        );
    }
```

File: packages/plugin/src/Bundles/Submissions/CipherFieldsBundle.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Bundles/Submissions/CipherFieldsBundle.php]: We don't need to check for the user or submission access.
Also we should probably decrypt the values even for `Lite` editions.
Imagine a scenario where people roll back their edition only to find all of their fields encrypted and they're no longer able to read them.

File: packages/plugin/src/Elements/Db/SubmissionQuery.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Elements/Db/SubmissionQuery.php]: There's an `SubmissionQuery::EVENT_AFTER_POPULATE_ELEMENT` and `SubmissionQuery::EVENT_AFTER_POPULATE_ELEMENTS` events. I would rather have our encryption bundle listen to those and perform the encrypt/decrypt there.

File: packages/plugin/src/Elements/Submission.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Elements/Submission.php]: Same here, there's a `Submission::EVENT_BEFORE_SAVE` event, which we can listen to, to perform encryption.

File: packages/plugin/src/Fields/AbstractField.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Fields/AbstractField.php]: `4` will definitely not work well with fields that have more than 4 sections.
It's `999` because it should always be at the very bottom.

File: packages/plugin/src/Fields/AbstractField.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Fields/AbstractField.php]: if we would use `protected bool $encrypted`, we could call the getter `isEncrypted()` instead. (Otherwise `PropertyAccessor` won't find this property.)

File: packages/plugin/src/Services/SubmissionsService.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Services/SubmissionsService.php:290]: If you use `HTMLField` or any other field which doesn't store values in the DB, you will get a critical failure.

What's the idea behind this change?

File: packages/plugin/src/templates/_components/modals/export_csv.html
Comment (gustavs-gutmanis) [inline packages/plugin/src/templates/_components/modals/export_csv.html]: Why this change?
I guess it should have been `{{ form.id == selectedFormId ? 'selected' }}`

File: packages/plugin/src/templates/_components/modals/export_csv.html
Comment (gustavs-gutmanis) [inline packages/plugin/src/templates/_components/modals/export_csv.html]: `??` makes `checked` output only when `isChecked` is `null`.
`?` is the first half of a terniary expression, which means it will output `checked` if `isChecked` evaluates to `true`.

It is a shorthand for
`{{ isChecked ? 'checked' : '' }}`
Twig lets us not write the `: ''` part, since it's useless here.

File: packages/plugin/src/controllers/export/QuickExportController.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/controllers/export/QuickExportController.php]: Since we have the `Field` instance a couple of lines back, we can know if the value uses encryption or not, so we could avoid calling `decrypt` on all values.

It might be redundant, though, but this logic is used 3 times across the codebase and would be a prime candidate for being extracted somewhere, so as not to duplicate it.

File: packages/plugin/src/Fields/Interfaces/NoEncryptionInterface.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Fields/Interfaces/NoEncryptionInterface.php]: Lets make this into an `EncryptionInterface` and have it have just `isEncrypted()` method.

Then make an `EncryptionTrait` which holds the property and its definition as well as the getter.

Then we assign it to the fields which can be encrypted, like `Text`, `Options` etc. 
Remember - there are several base fields which are extended by lots of other fields, so we need to apply the interface and trait only to the base ones. (i.e. `Text` will cover a wast amount of other fields).

File: packages/plugin/src/Fields/AbstractField.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Fields/AbstractField.php]: Advanced sections should go below the `Attributes` section, so an order of `1000` here, I guess.

File: packages/plugin/src/Bundles/Fields/Types/FieldTypesProvider.php
Comment (gustavs-gutmanis) [reply packages/plugin/src/Bundles/Fields/Types/FieldTypesProvider.php]: After reviewing the whole PR I see that this is now a manual entry everywhere instead of a generated default one.

Why this approach?
Before - we made sure there's always a general default section where every property that doesn't belong to a specific section falls in.
Now it's not only crazy amounts of work for you to manually type it everywhere, but it also introduces a possibility for human error - simply forgetting to enter a section.
If you don't enter one for a property, it will not show up anywhere in the builder anymore.

I'm guessing that you removed this because of the `advanced` section? If that's the case, we should think of a way to handle that case automatically, rather than with this manual approach.

File: packages/client/src/app/pages/forms/edit/builder/tabs/layout/property-editor/editors/fields/field-properties.tsx
Comment (gustavs-gutmanis) [inline packages/client/src/app/pages/forms/edit/builder/tabs/layout/property-editor/editors/fields/field-properties.tsx]: I'm thinking that this won't work if there's no `encrypt` property.
This has to be revised on my end.
For now leave it as it is here.

File: packages/client/src/app/components/form-controls/index.tsx
Comment (gustavs-gutmanis) [inline packages/client/src/app/components/form-controls/index.tsx]: This should be covered already by the `PropertyProvider`.
It turns all `pro` fields into hidden ones if the edition isn't PRO.

File: packages/plugin/src/Bundles/Submissions/CipherFieldsBundle.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Bundles/Submissions/CipherFieldsBundle.php]: We should be checking for fields which have implemented the `EncryptionInterface` (which we have to make).

```php
$fields = $submission->getFieldCollection()->getList(EncryptionInterface::class);
```

And check if `$field->isEncrypted()` -> then do encryption

File: packages/plugin/src/Events/Submissions/CipherEvent.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Events/Submissions/CipherEvent.php]: We can avoid this event it seems.

File: packages/plugin/src/Fields/AbstractField.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Fields/AbstractField.php]: We can crate an `EncryptionTrait` which has the definition of this property, but call it `$encrypted`, for better naming of the getter method. (helps with `PropertyAccessor`)

File: packages/plugin/src/Fields/AbstractField.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Fields/AbstractField.php]: Lets go the inclusion route, rather than the exclusion one. Then we won't need this at all.

File: packages/plugin/src/Fields/BaseGeneratedOptionsField.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Fields/BaseGeneratedOptionsField.php]: I have to say, this copy & paste approach here is a bad solution to a problem which doesn't exist.
Lets try and avoid such practices like the plague if at all possible.

File: packages/plugin/src/Fields/Implementations/Pro/GroupField.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Fields/Implementations/Pro/GroupField.php]: Lets add `EncryptionInterface` to all the fields which can be encrypted, rather than `NoEncryptionInterface` to the ones that can't.

There are some fields like Checkbox fields which values probably don't makes sense to encrypt, so we wouldn't be targeting all fields I guess.

File: packages/plugin/src/Fields/Implementations/Pro/GroupField.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Fields/Implementations/Pro/GroupField.php]: Can be removed

File: packages/plugin/src/controllers/export/ProfilesController.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/controllers/export/ProfilesController.php]: This logic tends to repeat in several places, it would be nice to extract it into a single point of entry, so that we duplicate code as minimally as possible.

**Review by gustavs-gutmanis — COMMENTED** [review]
Adding some commentary.

File: packages/client/src/app/components/form-controls/index.tsx
Comment (seandelaney) [reply packages/client/src/app/components/form-controls/index.tsx]: I didn't know this.

File: packages/plugin/src/Services/SubmissionsService.php
Comment (seandelaney) [reply packages/plugin/src/Services/SubmissionsService.php:290]: ~`getStorableFields` doesn't exist anymore~

It does here. `$form->getLayout()->getFields()->getStorableFields()`

Updated

File: packages/plugin/src/templates/_components/modals/export_csv.html
Comment (seandelaney) [reply packages/plugin/src/templates/_components/modals/export_csv.html]: I was getting a template error. Updated to `{{ form.id == selectedFormId ? 'selected' }}`

File: packages/client/src/app/components/form-controls/index.tsx
Comment (gustavs-gutmanis) [reply packages/client/src/app/components/form-controls/index.tsx]: That's cool, just mentioning it, since I know it's easy to miss this stuff.

**Comment by seandelaney** [comment]
@kjmartens @gustavs-gutmanis 

I think we need to spend some time refactoring how exports work. Multiple export routes with multiple options.

**Comment by seandelaney** [comment]
@kjmartens 

> When I try viewing the submission data inside the CP, front end and "Updating Search Indexes" in Craft Queue, I get an error: `base64_decode(): Argument #1 ($string) must be of type string, array given`
> 
> When attempting to submit a multipage form (not sure if it's the fact that it's multipage or something else), the form fails at the final submit and has this error:
> 
> ```
> Exception 'TypeError' with message 'openssl_encrypt(): Argument #1 ($data) must be of type string, array given'
> 
> in /YOINK/freeform5.c4/vendor/yiisoft/yii2/base/Security.php:218
> ```
> 
> On a single page form it appeared to work ok for submit.

I cannot replicate these issues in latest changes.  Can you please test again.

**Review by kjmartens — CHANGES_REQUESTED** [review]
Array type fields seem to have issues with this feature. Affected fields are:

- Checkboxes
- Multiple Select
- File Upload
- File Upload Drag & Drop

1. When the form contains ANY fields that contain an array of options, when submitting the form (and that field has data), the form submit fails with an error:
`Exception 'TypeError' with message 'openssl_encrypt(): Argument #1 ($data) must be of type string, array given'  in /PATH/vendor/yiisoft/yii2/base/Security.php:218`
2. When the form contains ANY fields that contain an array of field data, it errors when loading the submission, even if it is NOT set to be encrypted.
```TypeError base64_decode(): Argument #1 ($string) must be of type string, array given```

**Review by kjmartens — DISMISSED** [review]

**Review by kjmartens — CHANGES_REQUESTED** [review]
Much closer to working correctly now. However... 😅 

- Exporting via Quick Export and Export Profiles does not correctly decrypt the array fields (Checkboxes, Multiple Select, File Upload, File Upload Drag & Drop.

**Review by kjmartens — APPROVED** [review]

**Comment by kjmartens** [comment]
Seems good now. We should just wait on @gustavs-gutmanis's review when he gets back. 🙂

**Review by kjmartens — DISMISSED** [review]
A couple more issues I found with editions:

1. Encryption is only supposed to be available for Pro, but appears on Lite as well (correctly hidden on Express).
2. Decryption should happen regardless of edition. For example, if I was using Pro and encrypted some data and then later on downgraded to Lite (or Express) edition, I should still be able to access/view the decrypted submission data. Currently, it shows old data as encrypted still.

**Review by kjmartens — CHANGES_REQUESTED** [review]
Latest Changes:

- Encryption still shows for `Lite` when it should only appear for Pro (correctly hidden on Express).

**Review by kjmartens — APPROVED** [review]

File: packages/plugin/src/Fields/Implementations/Pro/PasswordField.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Fields/Implementations/Pro/PasswordField.php:30]: Does this need `#[Input\Hidden]`?
Does react need to know about its hidden state somehow?

If not, I think we can remove the attribute.

File: packages/plugin/src/Library/Helpers/EncryptionHelper.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Library/Helpers/EncryptionHelper.php]: Why are we iterating over references here?
We don't seem to be changing the original at all.

[Here's an explanation on why you should call `unset()`](https://www.php.net/manual/en/control-structures.foreach.php) on the value after a loop that uses a reference value.

Comment (gustavs-gutmanis) [reply packages/plugin/src/Library/Helpers/EncryptionHelper.php]: Also - maybe it makes sense to `json_encode` the values before encrypting and `json_decode` them on decrypt?

File: packages/plugin/src/Library/Helpers/EncryptionHelper.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Library/Helpers/EncryptionHelper.php]: I would advise against using reference values when looping, it's easy to miss that breaking thing I linked to you previously.

File: packages/plugin/src/Library/Helpers/EncryptionHelper.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/Library/Helpers/EncryptionHelper.php]: I think we can safely do `string $value` in the arguments, since it must always be a string anyway.

Or better yet, keep it as `mixed` and do `json_encode` here for the value, and adjust the decrypt accordingly.

File: packages/plugin/src/controllers/SubmissionsController.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/controllers/SubmissionsController.php]: It seems like this variable is redundant, because it will always be defined in line 93.

File: packages/plugin/src/controllers/export/ProfilesController.php
Comment (gustavs-gutmanis) [inline packages/plugin/src/controllers/export/ProfilesController.php]: Keeping these PHPDoc tags correct is an impossible task, because each time something in the chain changes, they become obsolete or outdated, so there's no real benefit of writing them. I would just disable the specific inspection in your JetBrains IDE.

**Review by gustavs-gutmanis — COMMENTED** [review]
Looking real nice now,

Just a couple of more questions before the approval.

File: packages/plugin/src/Fields/Implementations/Pro/PasswordField.php
Comment (seandelaney) [reply packages/plugin/src/Fields/Implementations/Pro/PasswordField.php:30]: @gustavs-gutmanis This is hiding the Encrypt field data label/checkbox. Because Password extends Text, Encrypt field data was still showing as 1) its within Pro edition and 2) inheritances EncryptionInterface from Text (which cannot be removed).

Comment (gustavs-gutmanis) [reply packages/plugin/src/Fields/Implementations/Pro/PasswordField.php:30]: Yeah I thought so, but if you override the variable definition, woudn't the attribute be unset as well?
If not, than this is the way to go, for sure, but I was just wondering.

File: packages/plugin/src/Library/Helpers/EncryptionHelper.php
Comment (seandelaney) [reply packages/plugin/src/Library/Helpers/EncryptionHelper.php]: We need to loop over certain field values as the field types such as table are saving values as an array in the DB itself. We cannot json_encode/json_decode the whole value for these multiple value fields as then we'd need to refactor how these field values are saving/used, with a ripple effect across the whole app.

File: packages/plugin/src/Library/Helpers/EncryptionHelper.php
Comment (seandelaney) [reply packages/plugin/src/Library/Helpers/EncryptionHelper.php]: Its to allow field types like table, checkboxes etc that save values as arrays to work.

Comment (seandelaney) [reply packages/plugin/src/Library/Helpers/EncryptionHelper.php]: It cant be a string as then tables etc wont work.

File: packages/plugin/src/Library/Helpers/EncryptionHelper.php
Comment (gustavs-gutmanis) [reply packages/plugin/src/Library/Helpers/EncryptionHelper.php]: The DB and checkbox fields store their array values as a JSON string in the DB, it is converted to and from arrays when loading the data via records, etc.

And the encrypted value is a string anyway, so maybe it is possible?
If a value is encrypted, it's 100% an encoded JSON. So even a single string value would be JSON encoded as `"somevalue"` when decrypted.

File: packages/plugin/src/Library/Helpers/EncryptionHelper.php
Comment (seandelaney) [reply packages/plugin/src/Library/Helpers/EncryptionHelper.php]: Would you create a new array instead ?

Comment (gustavs-gutmanis) [reply packages/plugin/src/Library/Helpers/EncryptionHelper.php]: Good question.
I guess this would conserve more memory on larger datasets.
Lets keep it like this.

A quick side question - is each value encrypted and decrypted separately in an array dataset?

File: packages/plugin/src/Fields/Implementations/Pro/PasswordField.php
Comment (seandelaney) [reply packages/plugin/src/Fields/Implementations/Pro/PasswordField.php:30]: TBH, I'm not 100% happy with this as a solution. Setting encrypted to false alone doesn't hide it. Unless you know why, a dev would be very confused. What do you think about making Password not extend Text?

**Comment by gustavs-gutmanis** [comment]
I guess it's better to have this attribute than to get rid of the inheritance.

--
Gustavs Gutmanis

**Review by gustavs-gutmanis — APPROVED** [review]
Looks great, thanks!

**Pull request merged**
